Modules/_functoolsmodule.c

```diff
@@ -1330,8 +1365,7 @@ lru_cache_tp_traverse(lru_cache_object *self, visitproc visit, void *arg)
     lru_list_elem *link = self->root.next;
     while (link != &self->root) {
         lru_list_elem *next = link->next;
-        Py_VISIT(link->key);
-        Py_VISIT(link->result);
+        Py_VISIT(link);
```

Comment:
If we need to visit type object during GC, we can do it here. No need to make lru_list_elem instances to GC type.

```c
        Py_VISIT(link-&gt;key);
        Py_VISIT(link-&gt;result);
        Py_VISIT(Py_TYPE(link));
```

Reply:
I'll do a benchmark first, to see if it makes a difference.

Reply:
I've applied your optimisation in 45469a74f534136d9aabd748638e74a30af0f42d. Thank you, @methane!

Reply:
`Py_VISIT(link)` is no-op. You should not remove this:

```
        Py_VISIT(link->key);
        Py_VISIT(link->result);
```

Reply:
Of course! Thanks.

Reply:
I think we need `Py_VISIT(Py_TYPE(link));` because each `lru_list_elem` instances have a strong reference to `lru_list_elem_type`.

Reply:
Is there a way to test if that's needed? I tried the following:

```
$ cat gc.py 
import sys
import gc
N_ITER=10
for i in range(N_ITER):
    before = sys.gettotalrefcount()
    import _functools
    del sys.modules['_functools']
    del _functools
    gc.collect()
    after = sys.gettotalrefcount()
    print(f"{before=}, {after=}")
$ ./python.exe gc.py  # without Py_VISIT(Py_TYPE(link))
before=50116, after=50122
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
$ ./python.exe gc.py  # with Py_VISIT(Py_TYPE(link))
before=50116, after=50122
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
before=50126, after=50126
```

Reply:
How about this?

```
import sys
import gc

def f(n):
    return n*2

N_ITER=10

for i in range(N_ITER):
    before = sys.gettotalrefcount()
    import _functools
    w = _functools._lru_cache_wrapper(f, 42, False, tuple)
    w(1) # create reference: _lru_cache_wrapper -> list_elem -> list_elem_type
    _functools._lru_list_elem.xxx = _functools  # list_elem_type -> _functools
    _functools.w = w # _functools -> _lru_cache_wrapper
    # circular reference: _functools -> w -> elem instance -> elem_type -> _functools
    del w
    del _functools
    del sys.modules['_functools']
    gc.collect()
    after = sys.gettotalrefcount()
    print(f"{before=}, {after=}")
```

In this example, I used `_functools._lru_list_elem.xxx = _functools` to create circular reference.
But it is not allowed in Python 3.9. `_functools` module didn't export `_lru_list_elem` in Python 3.9.
I think this is a small regression.  It is internal type and it must not be exposed to Python.

Reply:
Thank you for your insight, @methane. This indeed exposed the leak:

```
$ ./python.exe ft.py  # without explicit visit
before=50282, after=50561
before=50565, after=50822
before=50822, after=51079
before=51079, after=51336
before=51336, after=51593
before=51593, after=51850
before=51850, after=52107
before=52107, after=52364
before=52364, after=52621
before=52621, after=52878
$ ./python.exe ft.py  # with explicit visit
before=50282, after=50288
before=50292, after=50292
before=50292, after=50292
before=50292, after=50292
before=50292, after=50292
before=50292, after=50292
before=50292, after=50292
before=50292, after=50292
before=50292, after=50292
before=50292, after=50292
```

Reply:
See my review below: IMO functools._lru_list_elem type must implement the GC protocol (traverse and clear).